File: golang/testsuite/networks_impl/test_network.go
/*
 * Copyright (c) 2020 - present Kurtosis Technologies LLC.
 * All Rights Reserved.
 */

package networks_impl

import (
	"encoding/json"
	"github.com/kurtosis-tech/example-microservice/api/api_service_client"
	"github.com/kurtosis-tech/example-microservice/datastore/datastore_service_client"
	"github.com/kurtosis-tech/kurtosis-client/golang/networks"
	"github.com/kurtosis-tech/kurtosis-client/golang/services"
	"github.com/kurtosis-tech/kurtosis-libs/golang/testsuite/services_impl/datastore"
	"github.com/palantir/stacktrace"
	"github.com/sirupsen/logrus"
	"os"
	"strconv"
)

const (
	datastoreServiceId services.ServiceID = "datastore"
	apiServiceIdPrefix                    = "api-"

	waitForStartupDelayMilliseconds = 1000
	waitForStartupMaxNumPolls       = 15
)

//  A custom Network implementation is intended to make test-writing easier by wrapping low-level
//    NetworkContext calls with custom higher-level business logic
type TestNetwork struct {
	networkCtx                *networks.NetworkContext
	datastoreServiceImage     string
	apiServiceImage           string
	datastoreClient           *datastore_service_client.DatastoreClient
	personModifyingApiClient  *api_service_client.APIClient
	personRetrievingApiClient *api_service_client.APIClient
	nextApiServiceId          int
}

func NewTestNetwork(networkCtx *networks.NetworkContext, datastoreServiceImage string, apiServiceImage string) *TestNetwork {
	return &TestNetwork{
		networkCtx:                networkCtx,
		datastoreServiceImage:     datastoreServiceImage,
		apiServiceImage:           apiServiceImage,
		datastoreClient:           nil,
		personModifyingApiClient:  nil,
		personRetrievingApiClient: nil,
		nextApiServiceId:          0,
	}
}

//  Custom network implementations usually have a "setup" method (possibly parameterized) that is used
//   in the Test.Setup function of each test
func (network *TestNetwork) SetupDatastoreAndTwoApis() error {

	if network.datastoreClient != nil {
		return stacktrace.NewError("Cannot add datastore client to network; datastore client already exists!")
	}

	if network.personModifyingApiClient != nil || network.personRetrievingApiClient != nil {
		return stacktrace.NewError("Cannot add API services to network; one or more API services already exists")
	}

	containerCreationConfig := services.NewContainerCreationConfigBuilder(
		"kurtosistech/example-microservices_datastore",
		"/test-volume",
	).WithUsedPorts(
		map[string]bool{"1323/tcp": true},
	).Build()

	generateRunConfigFunc := func(ipAddr string, generatedFileFilepaths map[string]string, staticFileFilepaths map[services.StaticFileID]string) (*services.ContainerRunConfig, error) {
		return services.NewContainerRunConfigBuilder().Build(), nil
	}

	datastoreServiceContext, hostPortBindings, err := network.networkCtx.AddService(datastoreServiceId, containerCreationConfig, generateRunConfigFunc)
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred adding the datastore service")
	}

	datastoreClient := datastore_service_client.NewDatastoreClient(datastoreServiceContext.GetIPAddress(), datastore.Port)

	err = datastoreClient.WaitForHealthy(waitForStartupMaxNumPolls, waitForStartupDelayMilliseconds)
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred waiting for the datastore service to become available")
	}

	logrus.Infof("Added datastore service with host port bindings: %+v", hostPortBindings)

	network.datastoreClient = datastoreClient

	personModifyingApiClient, err := network.addApiService()
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred adding the person-modifying API client")
	}
	network.personModifyingApiClient = personModifyingApiClient

	personRetrievingApiClient, err := network.addApiService()
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred adding the person-retrieving API client")
	}
	network.personRetrievingApiClient = personRetrievingApiClient

	return nil
}

//  Custom network implementations will also usually have getters, to retrieve information about the
//   services created during setup
func (network *TestNetwork) GetPersonModifyingApiClient() (*api_service_client.APIClient, error) {
	if network.personModifyingApiClient == nil {
		return nil, stacktrace.NewError("No person-modifying API client exists")
	}
	return network.personModifyingApiClient, nil
}
func (network *TestNetwork) GetPersonRetrievingApiClient() (*api_service_client.APIClient, error) {
	if network.personRetrievingApiClient == nil {
		return nil, stacktrace.NewError("No person-retrieving API client exists")
	}
	return network.personRetrievingApiClient, nil
}

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
func (network *TestNetwork) addApiService() (*api_service_client.APIClient, error) {

	if network.datastoreClient == nil {
		return nil, stacktrace.NewError("Cannot add API service to network; no datastore client exists")
	}

	serviceIdStr := apiServiceIdPrefix + strconv.Itoa(network.nextApiServiceId)
	network.nextApiServiceId = network.nextApiServiceId + 1
	serviceId := services.ServiceID(serviceIdStr)

	configFileKey := "config-file"

	type config struct {
		DatastoreIp   string `json:"datastoreIp"`
		DatastorePort int    `json:"datastorePort"`
	}

	configInitializingFunc := func(fp *os.File) error {
		logrus.Debugf("Datastore IP: %v , port: %v", network.datastoreClient.IpAddr(), network.datastoreClient.Port())
		configObj := config{
			DatastoreIp:   network.datastoreClient.IpAddr(),
			DatastorePort: network.datastoreClient.Port(),
		}
		configBytes, err := json.Marshal(configObj)
		if err != nil {
			return stacktrace.Propagate(err, "An error occurred serializing the config to JSON")
		}

		logrus.Debugf("API config JSON: %v", string(configBytes))

		if _, err := fp.Write(configBytes); err != nil {
			return stacktrace.Propagate(err, "An error occurred writing the serialized config JSON to file")
		}

		return nil
	}

	apiServiceContainerCreationConfig := services.NewContainerCreationConfigBuilder(
		"kurtosistech/example-microservices_api",
		"/test-volume",
	).WithUsedPorts(
		map[string]bool{"2434/tcp": true},
	).WithGeneratedFiles(map[string]func(*os.File) error{
		configFileKey: configInitializingFunc,
	}).Build()

	apiServiceGenerateRunConfigFunc := func(ipAddr string, generatedFileFilepaths map[string]string, staticFileFilepaths map[services.StaticFileID]string) (*services.ContainerRunConfig, error) {
		configFilepath, found := generatedFileFilepaths[configFileKey]
		if !found {
			return nil, stacktrace.NewError("No filepath found for config file key '%v'", configFileKey)
		}
		startCmd := []string{
			"./api.bin",
			"--config",
			configFilepath,
		}
		result := services.NewContainerRunConfigBuilder().WithCmdOverride(startCmd).Build()
		return result, nil
	}

	apiServiceContext, hostPortBindings, err := network.networkCtx.AddService(serviceId, apiServiceContainerCreationConfig, apiServiceGenerateRunConfigFunc)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred adding the API service")
	}

	apiClient := api_service_client.NewAPIClient(apiServiceContext.GetIPAddress(), 2434)

	err = apiClient.WaitForHealthy(waitForStartupMaxNumPolls, waitForStartupDelayMilliseconds)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred waiting for the api service to become available")
	}

	logrus.Infof("Added API service with host port bindings: %+v", hostPortBindings)
	return apiClient, nil
}

File: rust/testsuite/src/networks_impl/test_network.rs
use std::{borrow::Borrow, rc::Rc, time::Duration};
use anyhow::{Context, Result, anyhow};

use kurtosis_rust_lib::{networks::{network::Network, network_context::NetworkContext}, services::service::ServiceId};

use crate::services_impl::{api::{api_service::ApiService, api_container_config_factory::ApiContainerConfigFactory}, datastore::{datastore_container_config_factory::DatastoreContainerConfigFactory, datastore_service::DatastoreService}};

const DATASTORE_SERVICE_ID_STR: &str = "datastore";
const API_SERVICE_ID_PREFIX: &str = "api-";

const WAIT_FOR_STARTUP_TIME_BETWEEN_POLLS: Duration = Duration::from_secs(1);
const WAIT_FOR_STARTUP_MAX_NUM_POLLS: u32 = 15;

// A custom Network implementation is intended to make test-writing easier by wrapping low-level
//   NetworkContext calls with custom higher-level business logic
pub struct TestNetwork {
	network_ctx: NetworkContext,
	datastore_service_image: String,
	api_service_image: String,
    datastore_service: Option<Rc<DatastoreService>>,
    person_modifying_api_service: Option<Rc<ApiService>>,
    person_retrieving_api_service: Option<Rc<ApiService>>,
    next_api_service_id: u32,
}

impl TestNetwork {
    pub fn new(network_ctx: NetworkContext, datastore_service_image: String, api_service_image: String) -> TestNetwork {
        return TestNetwork {
            network_ctx,
            datastore_service_image,
            api_service_image,
            datastore_service: None,
            person_modifying_api_service: None,
            person_retrieving_api_service: None,
            next_api_service_id: 0,
        };
    }

    // Custom network implementations usually have a "setup" method (possibly parameterized) that is used
    //  in the Test.Setup function of each test
    pub fn setup_datastore_and_two_api_services(&mut self) -> Result<()> {
        if self.datastore_service.is_some() {
            return Err(anyhow!(
                "Cannot add datastore service to network; datastore already exists!"
            ));
        }
        if self.person_modifying_api_service.is_some() || self.person_retrieving_api_service.is_some() {
            return Err(anyhow!(
                "Cannot add API services to network; one or more API services already exists"
            ));
        }

        let config_factory = DatastoreContainerConfigFactory::new(self.datastore_service_image.clone());
        let (service, host_port_bindings, checker) = self.network_ctx.add_service(&DATASTORE_SERVICE_ID_STR.to_owned(), &config_factory)
            .context("An error occurred adding the datastore service")?;
        checker.wait_for_startup(&WAIT_FOR_STARTUP_TIME_BETWEEN_POLLS, WAIT_FOR_STARTUP_MAX_NUM_POLLS)
            .context("An error occurred waiting for the datastore service to start")?;
        info!("Added datastore service with host port bindings: {:?}", host_port_bindings);
        self.datastore_service = Some(service);

        let person_modifying_api_service = self.add_api_service()
            .context("An error occurred adding the person-modifying API service")?;
        self.person_modifying_api_service = Some(person_modifying_api_service);

        let person_retrieving_api_service = self.add_api_service()
            .context("An error occurred adding the person-retrieving API service")?;
        self.person_retrieving_api_service = Some(person_retrieving_api_service);

        return Ok(());
    }

    // Custom network implementations will also usually have getters, to retrieve information about the
    //  services created during setup
    pub fn get_person_modifying_api_service(&self) -> Result<&ApiService> {
        match &self.person_modifying_api_service {
            Some(service_box) => return Ok(service_box.borrow()),
            None => return Err(anyhow!(
                "No person-modifying API service exists"
            )),
        }
    }
    pub fn get_person_retrieving_api_service(&self) -> Result<&ApiService> {
        match &self.person_retrieving_api_service {
            Some(service_box) => return Ok(service_box.borrow()),
            None => return Err(anyhow!(
                "No person-retrieving API service exists"
            )),
        }
    }

    // ====================================================================================================
    //                                       Private helper functions
    // ====================================================================================================
    fn add_api_service(&mut self) -> Result<Rc<ApiService>> {
        let datastore;
        match &self.datastore_service {
            Some(service_box) => datastore = service_box,
            None => return Err(anyhow!(
                "Cannot add API service to network; no datastore service exists"
            )),
        }

        let service_id: ServiceId = format!("{}{}", API_SERVICE_ID_PREFIX, self.next_api_service_id);
        self.next_api_service_id += 1;

        let config_factory = ApiContainerConfigFactory::new(self.api_service_image.clone(), datastore.borrow());
        let (api_service, host_port_bindings, checker) = self.network_ctx.add_service(&service_id, &config_factory)
            .context("An error occurred adding the API service")?;
        checker.wait_for_startup(&WAIT_FOR_STARTUP_TIME_BETWEEN_POLLS, WAIT_FOR_STARTUP_MAX_NUM_POLLS)
            .context("An error occurred waiting for the API service to start")?;
        info!("Added API service with host port bindings: {:?}", host_port_bindings);
        return Ok(api_service)
    }
}

impl Network for TestNetwork {}